Add general support for PyTorch in Imperceptible ASR attack

hesseltuinhof opened the pull request:
# Description

This PR adds support to the framework-independent (PyTorch and TensorFlow) ASR attack.

Related to https://github.com/Trusted-AI/adversarial-robustness-toolbox/issues/273

**To-Do Checklist**:

* [ ] Test the attack with the PyTorch Deepspeech estimator
* [x] The `torch.stft` does not yet match the `tensorflow.signal.stft`. This needs to be identical so that both masking are identical
* [x] Move to `dev_1.5.0` base

## Type of change

Please check all the relevant options.

- [x] New feature (non-breaking)

# Testing

Attack comes with a set of tests.

**Test Configuration**:
- Fedora 31
- Python 3.8.5
- ART 1.5 development
- Torch 1.7.0

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

beat-buesser commented on art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:
```diff
@@ -58,11 +61,11 @@ class ImperceptibleASR(EvasionAttack):
     def __init__(
         self,
-        estimator: "TensorFlowV2Estimator",
+        estimator: Union["PyTorchEstimator", "TensorFlowV2Estimator"],
```
For typing we should create a new type in `art.utils` line 175 and call it "SPEECH_RECOGNIZER_TYPE" and make it a Union of `PyTorchDeepSpeech` and `TensorFlowLingvoAsr`.

beat-buesser commented on art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:
```diff
@@ -108,16 +107,34 @@ def __init__(
-        self._loss_gradient_masking_threshold_op_tf = self._loss_gradient_masking_threshold_tf(
-            self._delta, self._power_spectral_density_maximum_tf, self._masking_threshold_tf
-        )
+        if isinstance(self.estimator, TensorFlowV2Estimator):
```
I think to make this check reliable for custom estimators should we search for `TensorFlowV2Estimator` in `self.estimator.__mro__, e.g. if a user inherits from `TensorFlowLingvoAsr`?

beat-buesser commented on art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:
```diff
@@ -108,16 +107,34 @@ def __init__(
+                self._delta, self._power_spectral_density_maximum_tf, self._masking_threshold_tf
+            )
+
+        elif isinstance(self.estimator, PyTorchEstimator):
```
Similar comment as for the `if` above.

beat-buesser commented on art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:
```diff
@@ -381,13 +441,33 @@ def _approximate_power_spectral_density_tf(
+        psd_matrix_approximated = pow(10.0, 9.6) / torch.unsqueeze(psd_maximum_stabilized, 1) * psd_matrix
+
+        # return PSD matrix such that shape is (batch_size, window_size // 2 + 1, frame_length)
+        # return torch.transpose(psd_matrix_approximated, 1, 2)
```
Is this line code or comment?

hesseltuinhof replied:
Ups... Line 469 should have been removed :see_no_evil:

hesseltuinhof commented:
Thank you @beat-buesser for the review! I have applied the review changes and rebased the PR onto `dev_1.5.0`.

hesseltuinhof commented:
I have fixed the `torch.STFT` which was missing the correct `window` keyword. Now all STFT calls, whether `scipy.signal`, `tf.signal` or `torch`, return equal results, which is important to obtain the correct threshold masking loss :+1:

Furthermore, just while wrapping up PR #719 I introduced a bug by losing the line that updated the perturbation in stage 2. The current PR reverts the issue.
